### PR TITLE
Contributing guide: add Quickstart and test details

### PR DESCRIPTION
**Pull request by Pooja-Gangwar**
### Desired Outcome

This document would include a step-by-step series of instructions which walk a user through:

Loading policy required for using the Collection
Establishing a Conjur host for a remote node
Using the Conjur Variable plugin in playbooks to provide secrets to remote nodes
This process documented in a markdown document, which is linked to in the repo's CONTRIBUTING docs.

### Implemented Changes

As a developer, I want an easy-to-use development environment, so I can work on the collection, roles, plugins, and playbooks without needing to run the test suite. The repo should include scripts in bin/ to setup and teardown this env

### Connected Issue/Story

Resolves #[[ONYX-15260](https://ca-il-jira.il.cyber-ark.com:8443/browse/ONYX-15260)]

### Definition of Done
*At least 1 todo must be completed in the sections below for the PR to be
merged.*

#### Changelog

- [ ] The CHANGELOG has been updated, or
- [X] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [X] The changes in this PR do not require tests

#### Documentation

- [X] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [X] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [X] There are no security aspects to these changes

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,24 +1,178 @@
...
+For developers, it is easy-to-use development environment, so that they can work on the collection, roles, plugins, and playbooks without needing to run the test suite. The repo scripts is in dev/ to setup and teardown this environment.
+This docker-compose dev environment is really useful, including a few different services:
+-  A Conjur Open Source instance
+-  An Ansible control node
+-  Managed nodes to push tasks to
+
+The `dev` directory contains a `docker-compose` file which creates a development
+environment with a database container (`pg`, short for *postgres*), and a
+`conjur` server container with source code mounted into the directory
+`/cyberark/dev/`.
```
Comment (john-odonnell) [inline CONTRIBUTING.md:50]: I think we can consolidate this a bit:

```suggestion
The `dev` directory contains a `docker-compose` file which creates a development
environment with:
- a Conjur Open Source instance
- an Ansible control node
- a number of managed nodes
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,24 +1,178 @@
...
+   root@f75015718049:/cyberark/dev/#
+   ```
+
+   Once the `start` script finishes, you're in a Bash shell inside the Conjur
+   server container.  To
+
+   After starting Conjur, your instance will be configured with the following:
+   * Account: `cucumber`
+   * User: `admin`
+   * Password: Run `cat conjur.identity` inside the container shell to display the current logged-in identity (which is also the password)
```
Comment (john-odonnell) [inline CONTRIBUTING.md:71]: Some of this section isn't accurate:

```md
1. Start the environment:

   ```sh-session
   $ cd dev
   $ ./start.sh
   ...
   PLAY RECAP *********************************************************************
   ansibleplugingtestingconjurhostidentity-test_app_centos-1 : ok=17 ...
   ansibleplugingtestingconjurhostidentity-test_app_centos-2 : ok=17 ...
   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-1 : ok=16 ...
   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-2 : ok=16 ...
   $
   \```

   Once the `start` script finishes, the environment's managed nodes have each been
   configured with a Conjur identity.

   After starting Conjur, your instance will be configured with the following:
   * Account: `cucumber`
   * User: `admin`
   * Password: Run `conjurctl role retrieve-key cucumber:user:admin` inside the Conjur container shell to retrieve the admin user API key (which is also the password)
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,24 +1,178 @@
...
+
+1. Debug the server
+
+   ```sh-session
+   root@f39015718062:docker-compose exec <container-service-name> bash
+   <various startup messages, then finally:>
+   Use exit to stop
+   ```
```
Comment (john-odonnell) [inline CONTRIBUTING.md:79]: This section can be removed - debugging Conjur isn't a focus here.

**Review by john-odonnell — CHANGES_REQUESTED** [review]
Thanks for tackling this. Made some corrections/suggestions below.

**Review by john-odonnell — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,24 +1,162 @@
...
+   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-2 : ok=16 ...
+   
+   ```
+
+   After starting Conjur, your instance will be configured with the following:
+   * Account: `cucumber`
+   * User: `admin`
+   * Password: Run `conjurctl role retrieve-key cucumber:user:admin` inside the Conjur container shell to retrieve the admin user API key (which is also the  password)
+
+## Setup a Conjur OSS Environment
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Please share your suggestions on shared MS office doc , accordingly I will update here .

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,24 +1,178 @@
 # Contributing to the Ansible Conjur Collection
-
+Thanks for your interest in Conjur Collection . Before contributing, please take a moment to
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,24 +1,178 @@
...
+ ## Prerequisites
+
+To start developing and testing using our development scripts ,
+the following tools need to be installed:
+
+1. [Git][get-git] to manage source code
+2. [Docker][get-docker] to manage dependencies and runtime environments
+3. [Docker Compose][get-docker-compose] to orchestrate Docker environments
+4. [Ruby][install-ruby] - native installation 
+5. [Python][Python] - for lookup plug-in
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,24 +1,178 @@
...
+    annotations:
+      description: Password needed by the Ansible remote machine
+
+  - !permit
+    role: *remote_hosts_layer
+    privileges: [ execute ]
+    resources: [ !variable target-password ]
+```
+
+## Useful links
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,24 +1,178 @@
...
+
+| Source  | URLs |
+| ------ | ------ |
+| CyberArk Conjur |https://docs.conjur.org/Latest/en/Content/Integrations/ansible.html|
+| GitHub | https://github.com/cyberark/ansible-conjur-collection|
+| Ansible Galaxy | https://galaxy.ansible.com/cyberark/conjur_collection|
+| Ansible Doc| https://docs.ansible.com/ansible/latest/collections/cyberark/conjur/index.html|
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -34,3 +92,76 @@ To run all tests:
...
+  Policy defines Conjur entities and the relationships between them.  An entity can be a policy, a host, a user, a layer, a group, or a variable.
+
+  ```sh
+    docker exec conjur_client conjur policy load root /policy/root.yml
+  ```
+## Create Ansible managed nodes
+
+  Set all required variables, ensure `Conjur_host_factory_token` is set . Install `ca-certificates` and place Conjur public SSL certificate. Symlink Conjur public SSL certificate into /etc/ssl/certs .
+
+## Use conjur host identity to set up Conjur identity on managed nodes
```
Comment (john-odonnell) [inline CONTRIBUTING.md:155]: The content of this section should be replaced with the content of the section `### Granting a Conjur identity to Ansible hosts`. This section currently tells the user _how_ the role will configure their managed nodes - we should focus on usage instructions here.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -34,3 +92,76 @@ To run all tests:
...
+
+
+  - Set all required variables, ensure `Conjur_host_factory_token` is set . Install `ca-certificates` , place Conjur  public SSL certificate and Symlink Conjur public SSL certificate into /etc/ssl/certs
+  - Install openssl-perl Package when ansible_os_family is `RedHat`
+  - Copy files from the Ansible to the hosts  into /etc/Conjur.conf and request identity from Conjur and place Conjur identity  file into /etc/
+
+## Use conjur variable lookup plugin to provide secrets to Ansible Playbooks
```
Comment (john-odonnell) [inline CONTRIBUTING.md:162]: This section should include a sample playbook that uses the lookup plugin. Something basic like this should do:
```yaml
---
- hosts: testapp
  tasks:
  - name: Provide secret with Lookup plugin
    debug:
      msg: "{{ lookup('cyberark.conjur.conjur_variable', '/ansible/target-password') }}"
```

**Review by john-odonnell — CHANGES_REQUESTED** [review]
Left a few comments below, mainly:
- We should provide code samples for everything we ask a user to do
- A bit of reorganizing

File: CONTRIBUTING.md (hunk 10 above)
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md (hunk 9 above)
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -34,3 +92,76 @@ To run all tests:
...
+
+
+  Policy defines Conjur entities and the relationships between them.  An entity can be a policy, a host, a user, a layer, a group, or a variable.
+
+  ```sh
+    docker exec conjur_client conjur policy load root /policy/root.yml
+  ```
+## Create Ansible managed nodes
+
+  Set all required variables, ensure `Conjur_host_factory_token` is set . Install `ca-certificates` and place Conjur public SSL certificate. Symlink Conjur public SSL certificate into /etc/ssl/certs .
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -34,3 +92,76 @@ To run all tests:
...
+
+## Set up Conjur Open Source and Ansible control node
+  - ### Setup a conjur OSS Environment
+
+    -   Pull the Docker image
+    -   Generate the master key
+    -   Load master key as an environment variable
+    -   Start the Conjur OSS environment
+    -   Create admin account
+    -   Connect the Conjur client to the Conjur server
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -34,3 +92,76 @@ To run all tests:
...
+    - Create a PR with these changes
+
+- Create an annotated tag with the new version, formatted as `v##.##.##`
+    - This will kick off an automated script which publish the release to
+      [Ansible Galaxy](https://galaxy.ansible.com/cyberark/conjur)
+
+- Create the release on GitHub for that tag
+    - Build the release package with `./ci/build_release`
+    - Attach package to Github Release
+
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md:113]: Done

**Comment by Pooja-Gangwar** [comment]
> Left a few comments below, mainly:
> 
> * We should provide code samples for everything we ask a user to do
> * A bit of reorganizing

Sorry for your time and efforts . I agree code sample will make the things easier to understand .
But I think long back we were agree to not mention the script file commands that's why I avoided .

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,7 +1,119 @@
 # Contributing to the Ansible Conjur Collection
 
+Thanks for your interest in Conjur. Before contributing, please take a moment to
+read and sign our <a href="https://github.com/cyberark/community/blob/master/documents/CyberArk_Open_Source_Contributor_Agreement.pdf" download="conjur_contributor_agreement">Contributor Agreement</a>.
+This provides patent protection for all Conjur users and allows CyberArk to enforce
+its license terms. Please email a signed copy to <a href="oss@cyberark.com">oss@cyberark.com</a>.
 For general contribution and community guidelines, please see the [community repo](https://github.com/cyberark/community).
 
+- Contributing to the Ansible Conjur Collection
```
Comment (john-odonnell) [inline CONTRIBUTING.md:9]:
```suggestion
- [Contributing to the Ansible Conjur Collection](#contributing-to-the-ansible-conjur-collection)
```

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,7 +1,119 @@
...
+    2. [Load policy to set up Conjur Ansible integration](#load-policy-to-set-up-conjur-ansible-integration)
+    3. [Create Ansible managed nodes](#create-ansible-managed-nodes)
+    4. [Use `conjur_host_identity` to set up Conjur identity on managed nodes](#use-conjur-host-identity-to-set-up-conjur-identity-on-managed-nodes)
+    5. [Use `conjur_variable` lookup plugin to provide secrets to Ansible Playbooks](#use-conjur-variable-lookup-plugin-to-provide-secrets-to-ansible-playbooks)
+
+
+ ## Prerequisites
+
+To start developing and testing using our development scripts ,
+the following tools need to be installed:
```
Comment (john-odonnell) [inline CONTRIBUTING.md:27]:
```suggestion
## Prerequisites

Before getting started, the following tools need to be installed:
```

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -1,7 +1,119 @@
...
+environment :
+-  A Conjur Open Source instance
+-  An Ansible control node
+-  Managed nodes to push tasks to
+
+To use it:
+
+1. Install dependencies (as above)
+
+1. To setup the dev environment ,first need to Clone GitHub [conjur-collection](https://github.com/cyberark/ansible-conjur-collection) repository in your directory and then run start_oss.sh script
```
Comment (john-odonnell) [inline CONTRIBUTING.md:49]:
```suggestion
1. To use the dev environment, clone the [Collection repository](https://github.com/cyberark/ansible-conjur-collection) and run the setup script:
```

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -1,7 +1,119 @@
...
+
+ ```sh-session
+ $ git clone https://github.com/cyberark/ansible-conjur-collection.git
+ $ cd ansible-conjur-collection/dev
+ $ ./start.sh
+
+ ```
+### Verification
+
+  When start_oss.sh script successfully setup Conjur environment along with inventory machines , the terminal returns the following:
```
Comment (john-odonnell) [inline CONTRIBUTING.md:60]:
```suggestion
When the Conjur and Ansible containers have been successfully setup, the terminal prints the following:
```

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -1,7 +1,119 @@
...
+   ...
+   PLAY RECAP *********************************************************************
+   ansibleplugingtestingconjurhostidentity-test_app_centos-1 : ok=17 ...
+   ansibleplugingtestingconjurhostidentity-test_app_centos-2 : ok=17 ...
+   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-1 : ok=16 ...
+   ansibleplugingtestingconjurhostidentity-test_app_ubuntu-2 : ok=16 ...
+
+   ```
+
+   After starting Conjur, your instance will be configured with the following:
```
Comment (john-odonnell) [inline CONTRIBUTING.md:72]:
```suggestion
Your Conjur instance will be configured with the following:
```

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -1,7 +1,119 @@
...
+   * Account: `cucumber`
+   * User: `admin`
+   * Password: Run `conjurctl role retrieve-key cucumber:user:admin` inside the Conjur container shell to retrieve the admin user API key (which is also the  password)
```
Comment (john-odonnell) [inline CONTRIBUTING.md:75]:
```suggestion
   * Password: Run `conjurctl role retrieve-key cucumber:user:admin` inside the Conjur container shell to retrieve the admin user's API key
```

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -1,7 +1,119 @@
...
+
+### Useful links
+
+- [Official documentation for Conjur's Ansible integration](https://docs.conjur.org/Latest/en/Content/Integrations/ansible.html)
+- [Conjur Collection on Ansible Galaxy](https://galaxy.ansible.com/cyberark/conjur)
+- [Ansible documentation for the Conjur collection](https://docs.ansible.com/ansible/latest/collections/cyberark/conjur/index.html)
+
+## Testing Conjur OSS
```
Comment (john-odonnell) [inline CONTRIBUTING.md:83]: We should change the `Testing Conjur OSS`, `Testing Conjur Enterprise`, and `Generate conjur variable unit tests report` sections into these:

```
## Testing

### Unit tests

Unit tests are only available for the Conjur Variable Lookup plugin. To run these tests:

\```
./dev/test_unit.sh
\```

### Integration tests

The collection has integration tests for both the Variable Lookup plugin and the Host Identity role that will validate each against live Conjur and Ansible containers.

To run all tests:
\```
./ci/test.sh -a
\```

To run the tests for a particular module:
\```
./ci/test.sh -d <role or plugin name>
\```

Integration tests can be run against Conjur Enterprise by adding the `-e` flag:
\```
./ci/test/sh -e -a
\```
```

**Review by john-odonnell — CHANGES_REQUESTED** [review]
Left a few comments below

**Comment by Pooja-Gangwar** [comment]
> Left a few comments below

Thanks John , made all suggested changes .

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -20,56 +125,95 @@ of this plugin:
...
-     $ cd tests
-     $ ./test.sh -a
-     ```
+## Use conjur host identity to set up Conjur identity on managed nodes
+
+  To grant your Ansible host a Conjur identity, you first must install the Conjur Ansible Role in your playbook directly
+
+  ```sh-session
+  ansible-galaxy install cyberark.conjur-host-identity
+  ```
```
Comment (john-odonnell) [inline CONTRIBUTING.md:196]:
```suggestion
  To grant your Ansible host a Conjur identity, first install the Conjur Collection on your Ansible control node:

  \```sh-session
  ansible-galaxy collection install cyberark.conjur
  \```
```

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -20,56 +125,95 @@ of this plugin:
...
+  Once you've done this, you can configure each Ansible node with a Conjur identity by including a section like the example below in your Ansible playbook:
+
+  ```sh-session
+  - hosts: servers
+    roles:
+      - role: cyberark.conjur-host-identity
```
Comment (john-odonnell) [inline CONTRIBUTING.md:202]:
```suggestion
      - role: cyberark.conjur.conjur-host-identity
```

File: CONTRIBUTING.md [hunk 23]
```diff
@@ -20,56 +125,95 @@ of this plugin:
...
-     $ cd tests
-     $ ./test.sh -d <role or plugin name>
-     ```
-    To run all tests:
-
-     ```sh-session
-     $ cd tests
-     $ ./test.sh -a
-     ```
+## Use conjur host identity to set up Conjur identity on managed nodes
```
Comment (john-odonnell) [inline CONTRIBUTING.md:190]:
```suggestion
## Use Conjur Ansible Role to set up identity on managed nodes
```

File: CONTRIBUTING.md [hunk 24]
```diff
@@ -20,56 +125,95 @@ of this plugin:
...
+  ```
+  Once you've done this, you can configure each Ansible node with a Conjur identity by including a section like the example below in your Ansible playbook:
+
+  ```sh-session
+  - hosts: servers
+    roles:
+      - role: cyberark.conjur-host-identity
+        conjur_appliance_url: 'https://conjur.myorg.com',
+        conjur_account: 'myorg',
+        conjur_host_factory_token: "{{lookup('env', 'HFTOKEN')}}",
```
Comment (john-odonnell) [inline CONTRIBUTING.md:205]: We need an additional step before this one setting up the host factory token in the `HFTOKEN` env var

File: CONTRIBUTING.md [hunk 25]
```diff
@@ -20,56 +125,95 @@ of this plugin:
...
+        conjur_host_name: "{{inventory_hostname}}"
+  ```
+  First we register the host with Conjur, adding it into the layer specific to the provided host factory token, and then installs Summon with the Summon Conjur provider for secret retrieval from Conjur.
+
+## Use conjur variable lookup plugin to provide secrets to Ansible Playbooks
```
Comment (john-odonnell) [inline CONTRIBUTING.md:210]:
```suggestion
## Use Conjur Ansible Lookup plugin to provide secrets to Ansible Playbooks
```

**Review by john-odonnell — CHANGES_REQUESTED** [review]
A couple more comments regarding the quickstart, left them below:

File: CONTRIBUTING.md (hunk 25 above)
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md:210]: Done

File: CONTRIBUTING.md (hunk 24 above)
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md:205]: Done

File: CONTRIBUTING.md (hunk 22 above)
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md:202]: Done

File: CONTRIBUTING.md (hunk 23 above)
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md:190]: Done

File: CONTRIBUTING.md (hunk 21 above)
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md:196]: Done

File: CONTRIBUTING.md [hunk 26]
```diff
@@ -20,56 +125,95 @@ of this plugin:
...
-   $ ./start.sh
-   ...
-   root@f75015718049:/cyberark/dev/#
+  docker-compose exec conjur_cli cat /policy/root.yml
+  docker-compose exec conjur_cli conjur policy load root /policy/root.yml
   ```
 
-   Once the `start` script finishes, you're in a Bash shell inside the Conjur
-   server container.  To
+  Also, load a dummy secret value into the 'ansible/target-password' variable. This is a variable required by remote nodes in order to complete their workloads.
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md:169]: Done

File: CONTRIBUTING.md [hunk 27]
```diff
@@ -20,56 +125,95 @@ of this plugin:
...
+
+    ```sh-session
+    export CLI_CONJUR_AUTHN_API_KEY="$(docker-compose exec conjur conjurctl role retrieve-key cucumber:user:admin)"
+    ```
+
+- Start the Conjur CLI container. The CLI will be automatically authenticated as the user `cucumber:user:admin`.
+
+    ```sh-session
+    docker-compose up -d conjur-cli
+    ```
```
[the remark this answers is not in the document]
Comment (Pooja-Gangwar) [reply CONTRIBUTING.md:156]: Done

**Review by john-odonnell — APPROVED** [review]
LGTM 👍

**Pull request merged**